### DISCUSSION TITLE
check idle

### DISCUSSION THREAD

**Discussion by Maksim-Siniugin** [coder/code-server#5406] [Q&A] [answered]
I have several codeservers running, how can I find out if someone is using one of them, or is it idle?

jupyterlab has an api for this. For example /api/kernels will show the status of the kernels, whether they are active or idle, or /api/terminals will show the status of running terminals.

Maybe there is something similar in codeserver to explicitly notice user activity inside the server?

**Comment by jsjoeio** [comment, marked as answer]
I think you could check the /healthz endpoint

I believe it beats every second and allows you to indicate whether or not there is activity:
- https://github.com/coder/code-server/blob/main/src/node/heart.ts
- https://github.com/coder/code-server/blob/main/test/unit/node/heart.test.ts
- https://github.com/coder/code-server/blob/main/src/node/routes/index.ts#L58-L62

**Reply by Maksim-Siniugin** [reply]
/healthz endpoint shows whether the server is alive or dead, but I would like to see if there is user activity or not

**Reply by jsjoeio** [reply]
Oh apologies - I misunderstood. So you want to know if there's user activity on the client aka interactions inside VS Code?

Maybe there is something they have there: summoning @code-asher

**Reply by code-asher** [reply]
You could use the `status` property of `/healthz`.  This will be "alive" if there are active connections to code-server (in other words someone has a browser window open) in the last minute.  Or you could use `lastHeartbeat` which is the last time code-server had an active connection (updated every minute).

Others use the heartbeat file on disk in `~/.local/share/code-server/heartbeat`.  The timestamp updates every minute so that can be used to create an idle timer.

There is not currently a way to differentiate between someone actively using code-server versus just having the window open and doing nothing though.

https://github.com/coder/code-server/blob/main/docs/FAQ.md#what-is-the-healthz-endpoint
https://github.com/coder/code-server/blob/main/docs/FAQ.md#what-is-the-heartbeat-file

**Reply by Maksim-Siniugin** [reply]
ok thx

**Reply by spencerng94** [reply]
How does the server determine if it is alive or dead? 

"This will be "alive" if there are active connections to code-server (in other words someone has a browser window open) in the last minute."

How does it check for active connections though? Where in the code is this?

**Reply by code-asher** [reply]
It asks the HTTP server how many things are connected to it:

https://github.com/coder/code-server/blob/5b62943216d7535255fd6081a452bd5204cf9b61/src/node/routes/index.ts#L34-L52

**Reply by code-asher** [reply]
Oh and by "things" specifically we are expecting at least two connections from the two web sockets VS Code opens (although there might be temporarily more connections for transient requests for assets and the like).
